design/metadata-handling.md

```diff
@@ -0,0 +1,286 @@
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Cloud-init templates offer a very powerful feature to render the configuration
+at runtime based on metadata given by Ironic. We want to take advantage of this
+feature to render node specific configuration for nodes in a machine deployment
+or in a control plane kubeadm object. This design proposal introduces a new
+metadata object that contains a list of items to provide metadata to Ironic for
+each node of a machine deployment or control plane.
+
+This proposal introduces a new metadata field in BareMetalHost and in
+Metal3Machine, and a metadataTemplate field in Metal3machine that allows
+generating the metadata content. It also introduces a new Metadata object that
+stores the list of metadata items for each node.
+
+## Motivation
+
+When deploying a machine deployment or a control plane with CAPI (v1alpha3), the
+KubeadmConfig applied is the same for all the nodes. However, we would want
+control on some fields, such as the node name for example. This can be a
+variable in the cloud-init template, but it needs to be provided through Ironic
+as metadata. And the metadata needs to be different for each of the nodes.
+For example we would want the nodes of a machine deployment to have names like
+`worker-np1-0`, `worker-np1-1`, `worker-np1-2` etc. and those names to be
+unrelated with the BareMetalHost names (that might be `machine-xyz`, reflecting
+some hardware related information, not Kubernetes node information).
+Hence we want a way to generate a node-specific metadata for a machine
+deployment and pass it down to Ironic as metadata.
+
+### Goals
+
+The goals are :
+
+- to add a metadata field in BareMetalHost to have it configurable
+  for Ironic, and in the Metal3Machine for CAPM3 to be able to pass it down to
+  BareMetalHost.
+- introduce new objects containing information to get the metadata per node,
+  or generate it, and store the values in use to not duplicate them.
```

Comment:
When you say "new objects" do you mean custom resources or something else? If custom resources, are they new types or existing types?

Reply:
I meant Custom Resources, and they are new types.